Extensions: update semantic model APIs

jcouv opened the pull request:
Updates GetDeclaredSymbol, LookupSymbols and GetMemberGroup
Adds ExtensionParameter

A number of APIs are out-of-scope for this PR and will be in follow-ups: GetOperation, GetSpeculativeSymbolInfo, GetSpeculativeTypeInfo, GetAliasInfo, GetSpeculativeAliasInfo, IsAccessible, AnalyzeDataFlow.

Relates to test plan https://github.com/dotnet/roslyn/issues/76130
Fixes https://github.com/dotnet/roslyn/issues/77502

jcouv commented on src/Compilers/CSharp/Portable/SymbolDisplay/SymbolDisplayVisitor.Types.cs:
```diff
@@ -338,7 +338,7 @@ private void AddNameAndTypeArgumentsOrParameters(INamedTypeSymbol symbol)
                 {
                     var extensionIdentifier = underlyingTypeSymbol!.ExtensionName;
-                    Builder.Add(CreatePart(SymbolDisplayPartKind.ClassName, symbol, extensionIdentifier)); // PROTOTYPE revisit when displaying extensions in the IDE
+                    Builder.Add(CreatePart(SymbolDisplayPartKind.ClassName, symbol, extensionIdentifier));
```
:memo: Confirmed this is fine with Cyrus

jcouv commented on src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/MethodGroup.cs:
```diff
@@ -47,41 +47,15 @@ internal void PopulateWithExtensionMethods(
             this.PopulateHelper(receiverOpt, resultKind, error);
             this.IsExtensionMethodGroup = true;
 
-            PooledHashSet<MethodSymbol> implementationsToShadow = null;
```
:memo: this deduplication between skeleton and implementation method is now handled upstream, in `EnumerateAllExtensionMembersInSingleBinder`

jcouv commented:
@AlekseyTs @jjonescz for review. Thanks

jjonescz commented on src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs:
```diff
@@ -1399,7 +1399,7 @@ private void CheckModelAndSyntaxNodeToSpeculate(CSharpSyntaxNode syntax)
         /// <param name="name">The name of the symbol to find. If null is specified then symbols
         /// with any names are returned.</param>
-        /// <param name="includeReducedExtensionMethods">Consider (reduced) extension methods.</param>
+        /// <param name="includeReducedExtensionMethods">Consider extension members. Classic extension methods will be returned in reduced form.</param>
```
Should the parameter be also renamed?
```suggestion
        /// <param name="includeReducedExtensionMembers">Consider extension members. Classic extension methods will be returned in reduced form.</param>
```

jcouv replied:
Sadly, this name is part of the public API, so changing it would be a break for any consumers that use an explicit argument name. That's why I left it. I can mark it as PROTOTYPE and we can revisit as part of API discussion

jcouv replied:
Oops, sorry, you're talking about the `CSharpSemanticModel` one which is internal. Yes, I'll do that

AlekseyTs commented on src/Compilers/Core/Portable/PublicAPI.Unshipped.txt:
```diff
@@ -2,7 +2,7 @@ Microsoft.CodeAnalysis.IEventSymbol.IsPartialDefinition.get -> bool
+Microsoft.CodeAnalysis.ITypeSymbol.ExtensionParameter.get -> Microsoft.CodeAnalysis.IParameterSymbol?
 override Microsoft.CodeAnalysis.Diagnostics.AnalyzerFileReference.ToString() -> string!
-
 Microsoft.CodeAnalysis.ITypeSymbol.IsExtension.get -> bool
```
I intentionally left an empty line as a separator for extensions specific changes.  This makes merges between branches go much smother.

AlekseyTs commented on src/Compilers/Core/Portable/PublicAPI.Unshipped.txt:
```diff
@@ -2,7 +2,7 @@ Microsoft.CodeAnalysis.IEventSymbol.IsPartialDefinition.get -> bool
 Microsoft.CodeAnalysis.IEventSymbol.PartialDefinitionPart.get -> Microsoft.CodeAnalysis.IEventSymbol?
 Microsoft.CodeAnalysis.IEventSymbol.PartialImplementationPart.get -> Microsoft.CodeAnalysis.IEventSymbol?
 Microsoft.CodeAnalysis.IncrementalGeneratorPostInitializationContext.AddEmbeddedAttributeDefinition() -> void
+Microsoft.CodeAnalysis.ITypeSymbol.ExtensionParameter.get -> Microsoft.CodeAnalysis.IParameterSymbol?
```
In order to simplify merges between branches, consider moving this addition to the end of the file where all extensions specific changes are grouped.

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/Source/SourceNamedTypeSymbol_Extension.cs:
```diff
@@ -114,5 +114,89 @@ internal sealed override ParameterSymbol? ExtensionParameter
+                bool success = symbol.CheckConstraints(
+                    new ConstraintsHelper.CheckConstraintsArgs(compilation, conversions, includeNullability: false, NoLocation.Singleton, diagnostics: null, template: CompoundUseSiteInfo<AssemblySymbol>.Discarded),
+                    substitution, typeParams, typeArgs, diagnosticsBuilder, nullabilityDiagnosticsBuilderOpt: null,
+                    ref useSiteDiagnosticsBuilder, ignoreTypeConstraintsDependentOnTypeParametersOpt: null);
```
>useSiteDiagnosticsBuilder [](http://example.com/codeflow?start=24&length=25)

Do we need to free this builder?

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Binder_Lookup.cs:
```diff
@@ -177,34 +183,72 @@ protected void LookupMembersInternal(LookupResult result, NamespaceOrTypeSymbol
-            Debug.Assert(!receiverType.IsDynamic());
-            if (!receiverType.IsErrorType())
+        internal void EnumerateAllExtensionMembersInSingleBinder(ArrayBuilder<SingleLookupResult> result,
+            TypeSymbol receiverType, string? name, int arity, LookupOptions options, Binder originalBinder,
```
>TypeSymbol receiverType [](http://example.com/codeflow?start=12&length=23)

I think this parameter should be removed from this helper and and from all callers that simply propagate its value. One shouldn't be required to provide a receiver type in order to enumerate extensions.

jcouv replied:
I like that! Thanks

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Binder_Lookup.cs:
```diff
@@ -177,34 +183,72 @@ protected void LookupMembersInternal(LookupResult result, NamespaceOrTypeSymbol
+
+                        if (!candidate.IsStatic)
+                        {
+                            implementationsToShadow ??= PooledHashSet<MethodSymbol>.GetInstance();
```
>implementationsToShadow ??= PooledHashSet<MethodSymbol>.GetInstance(); [](http://example.com/codeflow?start=28&length=70)

It looks like this statement can be moved into the following `if` statement

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Binder_Lookup.cs:
```diff
@@ -177,34 +183,72 @@ protected void LookupMembersInternal(LookupResult result, NamespaceOrTypeSymbol
+                        SingleLookupResult resultOfThisMember = originalBinder.CheckViability(candidate, arity, options, null, diagnose: true, useSiteInfo: ref classicExtensionUseSiteInfo);
+                        result.Add(resultOfThisMember);
+
+                        if (!candidate.IsStatic)
```
>if (!candidate.IsStatic) [](http://example.com/codeflow?start=24&length=24)

It looks like this condition is redundant (taking the comment for the line 222 into consideration)

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Binder_Lookup.cs:
```diff
@@ -177,34 +183,72 @@ protected void LookupMembersInternal(LookupResult result, NamespaceOrTypeSymbol
-            Binder originalBinder, ref CompoundUseSiteInfo<AssemblySymbol> useSiteInfo)
+        private void LookupAllExtensionMembersInSingleBinder(LookupResult result, TypeSymbol receiverType,
+            string? name, int arity, LookupOptions options, Binder originalBinder,
+            ref CompoundUseSiteInfo<AssemblySymbol> classicExtensionUseSiteInfo)
```
>classicExtensionUseSiteInfo [](http://example.com/codeflow?start=52&length=27)

It looks like the name doesn't reflect the purpose of the parameter

AlekseyTs commented on src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs:
```diff
@@ -4623,84 +4632,166 @@ internal static ImmutableArray<MethodSymbol> GetReducedAndFilteredMethodGroupSym
-                return false;
-            }
+                var diagnosticsBuilder = ArrayBuilder<TypeParameterDiagnosticInfo>.GetInstance();
+                ArrayBuilder<TypeParameterDiagnosticInfo>? useSiteDiagnosticsBuilder = null;
```
>useSiteDiagnosticsBuilder [](http://example.com/codeflow?start=59&length=25)

Do we need to free this builder?

AlekseyTs commented on src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs:
```diff
@@ -4712,65 +4803,63 @@ private static void MergeReducedAndFilteredMethodGroupSymbol(
                 else
                 {
-                    extensionThisType = call.Arguments[0].Type;
+                    receiverType = call.Arguments[0].Type;
```
>receiverType = call.Arguments[0].Type; [](http://example.com/codeflow?start=20&length=38)

This doesn't look appropriate for new extensions. If this code path is never taken for them, consider adding an assert.

jcouv replied:
In a scenario where overload resolution fails, we construct a bad BoundCall (see `CreateBadCall` in `BindInvocationExpressionContinued`) using the method group information. 
Method groups that are marked as `IsExtensionMethodGroup` hold an extra argument for the receiver. This was necessary to make overload resolution work on both kinds of members in one go.
For such method groups, the BoundCall will be marked as `InvokedAsExtensionMethod`.

```
                return CreateBadCall(node, methodGroup.Name, invokedAsExtensionMethod && analyzedArguments.Arguments.Count > 0 && (object)methodGroup.Receiver == (object)analyzedArguments.Arguments[0] ? null : methodGroup.Receiver,
                    GetOriginalMethods(result), methodGroup.ResultKind, methodGroup.TypeArguments.ToImmutable(), analyzedArguments, invokedAsExtensionMethod: invokedAsExtensionMethod, isDelegate: ((object)delegateTypeOpt != null));
```

One possible way we could improve this a bit (ie. contain the weirdness) would be to replace new extension methods with their implementation counterpart when we call `CreateBadCall`.

AlekseyTs commented on src/Features/Core/Portable/MetadataAsSource/AbstractMetadataAsSourceService.WrappedNamedTypeSymbol.cs:
```diff
@@ -146,6 +146,8 @@ public ImmutableArray<SymbolDisplayPart> ToMinimalDisplayParts(SemanticModel sem
 
         public bool IsExtension => _symbol.IsExtension;
 
+        public IParameterSymbol ExtensionParameter => _symbol.ExtensionParameter;
```
>_symbol.ExtensionParameter [](http://example.com/codeflow?start=54&length=26)

This looks suspicious, a wrapper should probably have its own extension parameter symbol (perhaps a wrapper itself).

AlekseyTs commented:
>     private static void AssertSetStrictlyEqual(string[] expected, string[] actual)

Consider using a name that better reflects the purpose of the method. For me "strictly" doesn't have any specific meaning and is not associated with uniqueness of elements

---
Refers to: src/Compilers/CSharp/Test/Emit3/Semantics/ExtensionTests.cs:40 in 693d896. [](commit_id = 693d896a0487bd6c5f09140763961c448790744f, deletion_comment = False)

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs:
```diff
@@ -8694,12 +8692,26 @@ static MethodGroupResolution resolveMethods(
+                        else if (method.GetIsNewExtensionMember()
+                            && SourceNamedTypeSymbol.GetCompatibleSubstitutedMember(binder.Compilation, method, receiverType) == null)
+                        {
+                            inapplicable = true;
```
>inapplicable = true; [](http://example.com/codeflow?start=28&length=20)

Are we testing this code path?

jcouv replied:
Yes. The interesting test is `PropertyAccess_NotAmbiguousWithInapplicableMethod`

AlekseyTs commented on src/Features/Core/Portable/MetadataAsSource/AbstractMetadataAsSourceService.WrappedNamedTypeSymbol.cs:
```diff
@@ -146,6 +146,10 @@
         public bool IsExtension => _symbol.IsExtension;
 
+        // PROTOTYPE this may need an implementation
+        public IParameterSymbol ExtensionParameter 
```
>public IParameterSymbol ExtensionParameter  [](http://example.com/codeflow?start=8&length=43)

It looks like there is a formatting error around this declaration

AlekseyTs reviewed (approved):
LGTM (commit 6)

jjonescz commented on src/Compilers/Core/Portable/PublicAPI.Unshipped.txt:
```diff
@@ -6,3 +6,4 @@ override Microsoft.CodeAnalysis.Diagnostics.AnalyzerFileReference.ToString() ->
 
 Microsoft.CodeAnalysis.ITypeSymbol.IsExtension.get -> bool
 Microsoft.CodeAnalysis.TypeKind.Extension = 14 -> Microsoft.CodeAnalysis.TypeKind
+Microsoft.CodeAnalysis.ITypeSymbol.ExtensionParameter.get -> Microsoft.CodeAnalysis.IParameterSymbol?
```
Do we have public API approval or tracking issue?

jcouv replied:
For feature branch, we can proceed without API approval. I'm tracking this as part of list of work still remaining for the feature

jcouv commented:
@AlekseyTs @jjonescz  The PR hit an issue in the constraint check I added in commit 3. Due to time constraint, I backed off that functionality for now (left a PROTOTYPE comment for follow-up) and added the repro scenario.

jjonescz commented:
> hit an issue

What was the issue, like a crash in the compiler?

jcouv commented:
> What was the issue, like a crash in the compiler?

Hit an assertion in `TypeMap` constructor inside `checkConstraintsIncludingExtension`. Aleksey and I sat and looked at it more just now and found a fix. I'll restore the constraints check in this PR

jcouv commented:
Pushed the fix: when checking constraints, the type parameters should all be definitions, and we can grab the `.TypeSubstitution` from symbol instead of creating a new `TypeMap`.

AlekseyTs reviewed (approved):
LGTM (commit 12)

AlekseyTs reviewed (approved):
LGTM (commit 13)
